include/seqan3/io/alignment_file/input.hpp

```diff
@@ -0,0 +1,1215 @@
+                         ref_ids_t          const & ref_ids,
+                         ref_sequences_t    const & ref_sequences,
+                         selected_field_ids const & SEQAN3_DOXYGEN_ONLY(fields_tag) = selected_field_ids{}) :
+        primary_stream{new std::ifstream{filename, std::ios_base::in | std::ios::binary}, stream_deleter_default}
+    {
+        // open stream
+        if (!primary_stream->good())
+            throw file_open_error{"Could not open file for reading."};
+
+        // possibly add intermediate compression stream
+        secondary_stream = detail::make_secondary_istream(*primary_stream, filename);
+
+        // initialise format handler or throw if format is not found
+        detail::set_format(format, filename);
+
+        // initialize reference information
+        set_references(ref_ids, ref_sequences);
+
+        // buffer first record
+        read_next_record();
+    }
+
+    /*!\brief Construct from an existing stream and with specified format.
+     * \tparam    file_format     The format of the file in the stream,
+     *                            must model seqan3::AlignmentFileInputFormat.
+     * \tparam    ref_ids_t       The range type that stores the reference ids.
+     * \tparam    ref_sequences_t The range type that stores the reference information.
+     * \param[in] stream          The stream to operate on; must be derived of std::basic_istream.
+     * \param[in] ref_ids         A range containing the reference ids that correspond to the SAM/BAM file.
+     * \param[in] ref_sequences   A range containing the reference sequences that correspond to the SAM/BAM file.
+     * \param[in] format_tag      The file format tag.
+     * \param[in] fields_tag      A seqan3::fields tag. [optional]
+     *
+     * \details
+     *
+     * The reference information given by the ids (names) and sequences will be used to construct a proper alignment
+     * when reading in SAM or BAM files. If you are not interested in the full alignment, you do not need to specify
+     * those information.
+     *
+     * In addition to the stream, reference information and format, you may specify a custom seqan3::fields object
+     * (e.g. `seqan3::fields<seqan3::field::SEQ>{}`) which may be easier than
+     * defining all the template parameters.
+     *
+     * ### Decompression
+     *
+     * This constructor transparently applies a decompression stream on top of the stream in case
+     * it is detected as being compressed.
+     * See the section on \link io_compression compression and decompression \endlink for more information.
+     */
+    template <IStream2                  stream_t,
+              std::ranges::ForwardRange ref_ids_t,
+              std::ranges::ForwardRange ref_sequences_t,
+              AlignmentFileInputFormat  file_format>
+    //!\cond
+        requires std::Same<value_type_t<ref_sequences_t>, ref_sequence_type> &&
+                 std::Same<value_type_t<ref_ids_t>, ref_id_type>
```

Comment:
So the `alignment_type` depends on the `ref_sequence_type` because when giving the reference sequence information it will be stored in the alignment object. So if the traits type says `dna4` I cannot provide `dna5` sequence because I would run into problems when I copy them over... (And I did not want to make those template parameters of the input file too...)

Maybe the ref_id_type does not matter on second thought..

Reply:
and maybe common reference type instead of value_type

Reply:
ah, ok. 

> and maybe common reference type instead of value_type

I think common type instead of common reference. But maybe also just implicitly-convertible? These concepts always confuse me, we should make a guide and add it to the wiki.